### ISSUE TITLE
HTTPretty breaking other URLs

### ISSUE THREAD

**Issue by spulec** [gabrielfalcao/HTTPretty#65]
With the following test

```python
@httprettified
def test_httpretty_should_not_override_other_urls():
    u"HTTPretty shouldn't break other URLs"

    HTTPretty.register_uri(HTTPretty.GET, "http://github.com/foo",
                           body="<root><baz /</root>")

    response = requests.get('http://google.com/')
    expect(response.status_code).to.equal(200)
```

I get

```
  File "HTTPretty/tests/functional/test_requests.py", line 153, in test_httpretty_should_not_override_other_urls
    response = requests.get('http://google.com/')
  File "HTTPretty/lib/python2.7/site-packages/requests/api.py", line 55, in get
    return request('get', url, **kwargs)
  File "HTTPretty/lib/python2.7/site-packages/requests/api.py", line 44, in request
    return session.request(method=method, url=url, **kwargs)
  File "HTTPretty/lib/python2.7/site-packages/requests/sessions.py", line 289, in request
    history = [r for r in gen] if allow_redirects else []
  File "HTTPretty/lib/python2.7/site-packages/requests/sessions.py", line 133, in resolve_redirects
    proxies=proxies
  File "HTTPretty/lib/python2.7/site-packages/requests/sessions.py", line 279, in request
    resp = self.send(prep, stream=stream, timeout=timeout, verify=verify, cert=cert, proxies=proxies)
  File "HTTPretty/lib/python2.7/site-packages/requests/sessions.py", line 374, in send
    r = adapter.send(request, **kwargs)
  File "HTTPretty/lib/python2.7/site-packages/requests/adapters.py", line 222, in send
    r.content
  File "HTTPretty/lib/python2.7/site-packages/requests/models.py", line 550, in content
    self._content = bytes().join(self.iter_content(CONTENT_CHUNK_SIZE)) or bytes()
  File "HTTPretty/lib/python2.7/site-packages/requests/models.py", line 496, in generate
    chunk = self.raw.read(chunk_size)
  File "HTTPretty/lib/python2.7/site-packages/requests/packages/urllib3/response.py", line 148, in read
    return self._fp.read(amt)
  File "python2.7/httplib.py", line 541, in read
    return self._read_chunked(amt)
  File "python2.7/httplib.py", line 592, in _read_chunked
    value.append(self._safe_read(amt))
  File "python2.7/httplib.py", line 649, in _safe_read
    raise IncompleteRead(''.join(s), amt)
IncompleteRead: IncompleteRead(532 bytes read)
```

**Comment by gabrielfalcao** [comment]
This might have been fixed by @frankamp on #216

**Comment by porterjamesj** [comment]
I believe I'm still seeing this on 0.8.8.

**Comment by GitFree** [comment]
seeing this in both 0.8.8 and 0.8.9.
log shows (Caused by <class 'httplib.BadStatusLine'>: '')

**Comment by anentropic** [comment]
I am getting `Connection aborted.', BadStatusLine("''",)` error on non-HTTPretty url when `@httpretty.activate` is active

when I run this:
```
    @httpretty.activate
    def test_user_subscribed_to_broadcast(self):
        httpretty.register_uri(httpretty.POST, 'https://broadcast.test/broadcast/v1/user',
                               body='Hello', status=201)
```

I get `BadStatusLine("''",)` error in a PUT request to `http://localhost:9200/elasticsearch-index/user/1?refresh=true`

(it sounds a bit like #182 except I get it for a uri _not_ registered with HTTPretty)

the request succeeds if I remove the decorator

if I drop into ipdb I see that I do have `allow_net_connect=True`

httpretty==0.8.10

**Comment by anentropic** [comment]
looks like same issue as #194 #199 #187

**Comment by anentropic** [comment]
FWIW whatever they're doing differently... https://github.com/patrys/httmock works fine

**Comment by boazin** [comment]
Still happening... getting a 
`requests.exceptions.ConnectionError: ('Connection aborted.', BadStatusLine("''",))`
Even on `http://localhost:8000`

Related?

**Comment by gabrielfalcao** [comment]
@boazin is the `@httpretty.activate` the decorator not working ?

**Comment by boazin** [comment]
Haven't tried the decorator. I have a boolean I need to check in a config to decide if I use the mock or the real URL, so I can't use the decorator - need to manually call `.enable()` `.disable()`

**Comment by m1schka** [comment]
I wouldn't consider this issue as closed
Similar to @anentropic, I get a `BadStatusLine` from inside the boto package even though I don't mock those requests (because I'm using fakes3 for testing.
I'm also using enable/disable manually, because I'm using it in test setup to activate a special api mock, which can occur at any given point in the tests.

**Comment by sybrenstuvel** [comment]
I can confirm this issue on both Python 2.7 and 3.5. This is my test script:

    from __future__ import print_function
    
    import requests
    import httpretty
    
    httpretty.enable()
    
    print('Requesting Google.')
    print(requests.get('http://www.google.com/'))
    print('Done!')

On Python 2.7.6 it hangs on the `requests.get(...)` line:

    Requesting Google.
    ^CTraceback (most recent call last):
      File "check_httpretty.py", line 9, in <module>
        print(requests.get('http://www.google.com/'))
      File "/home/sybren/.virtualenvs/pillar/local/lib/python2.7/site-packages/requests/api.py", line 67, in get
        return request('get', url, params=params, **kwargs)
      File "/home/sybren/.virtualenvs/pillar/local/lib/python2.7/site-packages/requests/api.py", line 53, in request
        return session.request(method=method, url=url, **kwargs)
      File "/home/sybren/.virtualenvs/pillar/local/lib/python2.7/site-packages/requests/sessions.py", line 468, in request
        resp = self.send(prep, **send_kwargs)
      File "/home/sybren/.virtualenvs/pillar/local/lib/python2.7/site-packages/requests/sessions.py", line 576, in send
        r = adapter.send(request, **kwargs)
      File "/home/sybren/.virtualenvs/pillar/local/lib/python2.7/site-packages/requests/adapters.py", line 376, in send
        timeout=timeout
      File "/home/sybren/.virtualenvs/pillar/local/lib/python2.7/site-packages/requests/packages/urllib3/connectionpool.py", line 559, in urlopen
        body=body, headers=headers)
      File "/home/sybren/.virtualenvs/pillar/local/lib/python2.7/site-packages/requests/packages/urllib3/connectionpool.py", line 353, in _make_request
        conn.request(method, url, **httplib_request_kw)
      File "/usr/lib/python2.7/httplib.py", line 973, in request
        self._send_request(method, url, body, headers)
      File "/usr/lib/python2.7/httplib.py", line 1007, in _send_request
        self.endheaders(body)
      File "/usr/lib/python2.7/httplib.py", line 969, in endheaders
        self._send_output(message_body)
      File "/usr/lib/python2.7/httplib.py", line 829, in _send_output
        self.send(msg)
      File "/usr/lib/python2.7/httplib.py", line 805, in send
        self.sock.sendall(data)
      File "/home/sybren/.virtualenvs/pillar/local/lib/python2.7/site-packages/httpretty/core.py", line 456, in sendall
        self.real_sendall(data)
      File "/home/sybren/.virtualenvs/pillar/local/lib/python2.7/site-packages/httpretty/core.py", line 380, in real_sendall
        received = self.truesock.recv(self._bufsize)
    KeyboardInterrupt

On Python 3.5.1 it produces:

    Requesting Google.
    Traceback (most recent call last):
      File "/home/sybren/.virtualenvs/pillar-python-sdk/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py", line 559, in urlopen
        body=body, headers=headers)
      File "/home/sybren/.virtualenvs/pillar-python-sdk/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py", line 353, in _make_request
        conn.request(method, url, **httplib_request_kw)
      File "/opt/python3.5/lib/python3.5/http/client.py", line 1083, in request
        self._send_request(method, url, body, headers)
      File "/opt/python3.5/lib/python3.5/http/client.py", line 1128, in _send_request
        self.endheaders(body)
      File "/opt/python3.5/lib/python3.5/http/client.py", line 1079, in endheaders
        self._send_output(message_body)
      File "/opt/python3.5/lib/python3.5/http/client.py", line 911, in _send_output
        self.send(msg)
      File "/opt/python3.5/lib/python3.5/http/client.py", line 885, in send
        self.sock.sendall(data)
      File "/home/sybren/.virtualenvs/pillar-python-sdk/lib/python3.5/site-packages/httpretty/core.py", line 456, in sendall
        self.real_sendall(data)
      File "/home/sybren/.virtualenvs/pillar-python-sdk/lib/python3.5/site-packages/httpretty/core.py", line 372, in real_sendall
        self.truesock.connect(self._address)
    OSError: [Errno 9] Bad file descriptor
    
    During handling of the above exception, another exception occurred:
    
    Traceback (most recent call last):
      File "/home/sybren/.virtualenvs/pillar-python-sdk/lib/python3.5/site-packages/requests/adapters.py", line 376, in send
        timeout=timeout
      File "/home/sybren/.virtualenvs/pillar-python-sdk/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py", line 609, in urlopen
        _stacktrace=sys.exc_info()[2])
      File "/home/sybren/.virtualenvs/pillar-python-sdk/lib/python3.5/site-packages/requests/packages/urllib3/util/retry.py", line 247, in increment
        raise six.reraise(type(error), error, _stacktrace)
      File "/home/sybren/.virtualenvs/pillar-python-sdk/lib/python3.5/site-packages/requests/packages/urllib3/packages/six.py", line 309, in reraise
        raise value.with_traceback(tb)
      File "/home/sybren/.virtualenvs/pillar-python-sdk/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py", line 559, in urlopen
        body=body, headers=headers)
      File "/home/sybren/.virtualenvs/pillar-python-sdk/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py", line 353, in _make_request
        conn.request(method, url, **httplib_request_kw)
      File "/opt/python3.5/lib/python3.5/http/client.py", line 1083, in request
        self._send_request(method, url, body, headers)
      File "/opt/python3.5/lib/python3.5/http/client.py", line 1128, in _send_request
        self.endheaders(body)
      File "/opt/python3.5/lib/python3.5/http/client.py", line 1079, in endheaders
        self._send_output(message_body)
      File "/opt/python3.5/lib/python3.5/http/client.py", line 911, in _send_output
        self.send(msg)
      File "/opt/python3.5/lib/python3.5/http/client.py", line 885, in send
        self.sock.sendall(data)
      File "/home/sybren/.virtualenvs/pillar-python-sdk/lib/python3.5/site-packages/httpretty/core.py", line 456, in sendall
        self.real_sendall(data)
      File "/home/sybren/.virtualenvs/pillar-python-sdk/lib/python3.5/site-packages/httpretty/core.py", line 372, in real_sendall
        self.truesock.connect(self._address)
    requests.packages.urllib3.exceptions.ProtocolError: ('Connection aborted.', OSError(9, 'Bad file descriptor'))
    
    During handling of the above exception, another exception occurred:
    
    Traceback (most recent call last):
      File "check_httpretty.py", line 9, in <module>
        print(requests.get('http://www.google.com/'))
      File "/home/sybren/.virtualenvs/pillar-python-sdk/lib/python3.5/site-packages/requests/api.py", line 67, in get
        return request('get', url, params=params, **kwargs)
      File "/home/sybren/.virtualenvs/pillar-python-sdk/lib/python3.5/site-packages/requests/api.py", line 53, in request
        return session.request(method=method, url=url, **kwargs)
      File "/home/sybren/.virtualenvs/pillar-python-sdk/lib/python3.5/site-packages/requests/sessions.py", line 468, in request
        resp = self.send(prep, **send_kwargs)
      File "/home/sybren/.virtualenvs/pillar-python-sdk/lib/python3.5/site-packages/requests/sessions.py", line 576, in send
        r = adapter.send(request, **kwargs)
      File "/home/sybren/.virtualenvs/pillar-python-sdk/lib/python3.5/site-packages/requests/adapters.py", line 426, in send
        raise ConnectionError(err, request=request)
    requests.exceptions.ConnectionError: ('Connection aborted.', OSError(9, 'Bad file descriptor'))

This is all on Ubuntu 14.04.

**Comment by m1schka** [comment]
I went back to using https://github.com/getsentry/responses, which is doing a better job at what I need

**Comment by david-caro** [comment]
I'm having this same issue on fedora 25, with python 2.7.13, requests 2.13.1 and httpretty>0.6.4 (tried all the versions to 0.8.4, and since 0.6.4 they just hang). My test

```
python -c "import httpretty;import requests;httpretty.enable();print requests.get('http://google.com')"
```

So it seems that a change introduced on 0.6.5 broke it...

**Comment by david-caro** [comment]
Yep, definetly the EAGAIN 'fix' broke it for me: 1d6f94e (in the sense that right before that commit it still works, after it, it hangs and breaks)

**Comment by davidread** [comment]
Downgrading to httpretty 0.6.4 works for me on Linux, but I get a new error on OSX, related to the proxy.
```
mkvirtualenv -p python3.6 test6
pip install requests
pip install 'httpretty<0.6.5'
python -c "import httpretty;import requests;httpretty.enable();print(requests.get('http://google.com'))"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/dread/.virtualenvs/test6/lib/python3.6/site-packages/requests/api.py", line 72, in get
    return request('get', url, params=params, **kwargs)
  File "/Users/dread/.virtualenvs/test6/lib/python3.6/site-packages/requests/api.py", line 58, in request
    return session.request(method=method, url=url, **kwargs)
  File "/Users/dread/.virtualenvs/test6/lib/python3.6/site-packages/requests/sessions.py", line 499, in request
    prep.url, proxies, stream, verify, cert
  File "/Users/dread/.virtualenvs/test6/lib/python3.6/site-packages/requests/sessions.py", line 672, in merge_environment_settings
    env_proxies = get_environ_proxies(url, no_proxy=no_proxy)
  File "/Users/dread/.virtualenvs/test6/lib/python3.6/site-packages/requests/utils.py", line 692, in get_environ_proxies
    if should_bypass_proxies(url, no_proxy=no_proxy):
  File "/Users/dread/.virtualenvs/test6/lib/python3.6/site-packages/requests/utils.py", line 676, in should_bypass_proxies
    bypass = proxy_bypass(netloc)
  File "/usr/local/Cellar/python3/3.6.2/Frameworks/Python.framework/Versions/3.6/lib/python3.6/urllib/request.py", line 2612, in proxy_bypass
    return proxy_bypass_macosx_sysconf(host)
  File "/usr/local/Cellar/python3/3.6.2/Frameworks/Python.framework/Versions/3.6/lib/python3.6/urllib/request.py", line 2589, in proxy_bypass_macosx_sysconf
    return _proxy_bypass_macosx_sysconf(host, proxy_settings)
  File "/usr/local/Cellar/python3/3.6.2/Frameworks/Python.framework/Versions/3.6/lib/python3.6/urllib/request.py", line 2563, in _proxy_bypass_macosx_sysconf
    hostIP = ip2num(hostIP)
  File "/usr/local/Cellar/python3/3.6.2/Frameworks/Python.framework/Versions/3.6/lib/python3.6/urllib/request.py", line 2542, in ip2num
    parts = list(map(int, parts))
ValueError: invalid literal for int() with base 10: 'google'
```
